Fix: utils: Add 'sudo' only when there is a sudoer(bsc#1215549)

liangxin1300 opened the pull request:
backport #1247

nicholasyang2022 reviewed (changes requested):
It is preferred to avoid using split to parse cmdline arguments to avoid shell injection.

liangxin1300 commented:
> It is preferred to avoid using split to parse cmdline arguments to avoid shell injection.

Changed on all 3 PRs

Thanks for pointing this!

nicholasyang2022 commented:
This change does not work. If whitespace characters exist in `host`, `shlex.split` will still split it. To avoid shell injection, arguments should not be composed into a command line without validation. For this case, it is preferred to operate on the argument array instead of command line string.

liangxin1300 commented:
> This change does not work. If whitespace characters exist in `host`, `shlex.split` will still split it. To avoid shell injection, arguments should not be composed into a command line without validation. For this case, it is preferred to operate on the argument array instead of command line string.

Changed
Please take a look.
Thanks!

nicholasyang2022 reviewed (approved):
~~{"ssh", "true"} is not the same as {"ssh", "", "true"}. However, it seems that ssh ignore an empty arg and this does not cause any problem.~~

nicholasyang2022 commented on crmsh/utils.py:
```diff
@@ -185,7 +185,8 @@ def _get_user_of_host_from_config(host):
         args = ['ssh']
         args.extend(constants.SSH_OPTION_ARGS)
-        args.extend(['-o', 'BatchMode=yes', host, 'sudo', 'true'])
+        sudo_str = 'sudo ' if userdir.get_sudoer() else ''
+        args.extend(['-o', 'BatchMode=yes', host, sudo_str, 'true'])
```
1. When passing arguments as array of strings, space is not needed. Space is the separator of arguments in shell. See `init main(int argc, char **argv)` and `execve(1)`.
2. An empty argument is not the same as non-existing argument. `argv = {"true"}` and  `argv = {"", "true"}` is not the same.

```suggestion
        if userdir.get_sudoer():
            args.extend(['-o', 'BatchMode=yes', host, 'sudo', 'true'])
        else:
            args.extend(['-o', 'BatchMode=yes', host, 'true'])
```
